apps/emqx_durable_storage/src/emqx_ds_replication_layer.erl

```diff
@@ -135,6 +141,21 @@ add_generation(DB) ->
 update_db_config(DB, CreateOpts) ->
     emqx_ds_replication_layer_meta:update_db_config(DB, CreateOpts).
 
+-spec list_generations_with_lifetimes(emqx_ds:db()) ->
+    #{generation_rank() => emqx_ds:generation_info()}.
+list_generations_with_lifetimes(DB) ->
+    Shards = list_shards(DB),
+    Nodes = lists:usort(lists:map(fun(Shard) -> node_of_shard(DB, Shard) end, Shards)),
+    Results = emqx_ds_proto_v3:list_generations_with_lifetimes(Nodes, DB),
+    %% TODO: Handle errors?
+    Gens = [Gen || {ok, Gen} <- Results],
+    lists:foldl(fun maps:merge/2, #{}, Gens).
+
+-spec drop_generation(emqx_ds:db(), generation_rank()) -> ok | {error, _}.
+drop_generation(DB, {Shard, GenId}) ->
+    Node = node_of_shard(DB, Shard),
```

Comment:
This is good for now, but maybe add a TODO that explains that this operation must be propagated, in one way or another, to the entire replica set.